File: Tutorial4-7LayerLibrary/src/main/java/com/test/tutorial4_7layerlibrary/di/BaseAppComponent.java
package com.test.tutorial4_7layerlibrary.di;


import android.app.Application;

import com.test.tutorial4_7layerlibrary.BaseActivity;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

@Singleton
@Component(modules = BaseAppModule.class)
public interface BaseAppComponent {

    void inject(BaseActivity baseActivity);

    @Component.Builder
    interface Builder {

        @BindsInstance
        Builder application(Application application);

        BaseAppComponent build();

    }
}

File: Tutorial4-7LayerLibrary/src/main/java/com/test/tutorial4_7layerlibrary/di/BaseAppComponent.kt
package com.test.tutorial4_7layerlibrary.di

import android.app.Application
import android.content.SharedPreferences
import com.test.tutorial4_7layerlibrary.model.LibraryObject
import com.test.tutorial4_7layerlibrary.model.ToastMaker
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [SubComponentsModule::class, BaseAppModule::class])
interface BaseAppComponent {

    fun librarySubComponentBuilder(): LibraryActivitySubComponent.Builder


    /**
     * 🔥🔥🔥 This method is required to get this object from a class that uses this component
     * as dependent component
     */
    fun getSharedPreferences(): SharedPreferences

    /**
     * 🔥🔥🔥 This method is required to get this object from a class that uses this component
     * as dependent component
     */
    fun getToastMaker(): ToastMaker

    /**
     * 🔥🔥🔥 This method is required to get this object from a class that uses this component
     * as dependent component
     */
    fun getLibraryObject(): LibraryObject


    @Component.Builder
    interface Builder {
        @BindsInstance
        fun application(application: Application): Builder
        fun build(): BaseAppComponent
    }
}